code/iaas/model/src/main/java/io/cattle/platform/core/dao/impl/InstanceDaoImpl.java

```diff
@@ -176,6 +176,15 @@ public Instance getInstanceByUuidOrExternalId(Long accountId, String uuid, Strin
                 .where(SERVICE_EXPOSE_MAP.INSTANCE_ID.eq(instance.getId()))
                 .fetchInto(ServiceRecord.class);
     }
+    
+    public List<? extends Service> findServicesForInstanceId(Long instanceId){
+    		return create().select(SERVICE.fields())
+    				.from(SERVICE)
+    				.join(SERVICE_EXPOSE_MAP)
+    				.on(SERVICE_EXPOSE_MAP.SERVICE_ID.eq(SERVICE.ID))
```

Comment:
should check for removed being null on both service and serviceexposemap

Reply:
https://github.com/leodotcloud/cattle/blob/e27b2641a783d877d6480644606ff442cd5cc5d8/code/iaas/model/src/main/java/io/cattle/platform/core/dao/impl/InstanceDaoImpl.java#L171

should it be added for `findServicesFor` as well then?